R/morphemepiece.R

```diff
@@ -126,67 +126,176 @@ magrittr::`%>%`
+      } else {
+        start <- start + 1 # backward; hold end fixed
+      }
+    }
+    
+    if (is.na(cur_substr[[1]]) ) {
+      is_bad <-  TRUE #nocov
+      break           #nocov
+    }
+    
+    if (dir == 1) {
+      sub_tokens <- append(sub_tokens, cur_substr) # append to end
+      start <- end + 1 # rest of word, after taking piece off the front
+      keepgoing <- start <= wordlen
+    } else {
+      sub_tokens <- append(cur_substr, sub_tokens) # append to beginning
+      end <- start - 1 # rest of word, after taking piece off the end
+      keepgoing <- end >= 1
+    }
+  }
+  
+  if (is_bad) {
+    return(unk_token) #nocov
+  }
+  return(sub_tokens)
+}
 
-        this_breakdown <- unlist(
-            .pull_out_longest_piece(current_word, mp_vocab_nohash,
-                                    original_start = original_start,
-                                    original_end = original_end,
-                                    prefixes = prefixes)
-        )
-        if (!stringr::str_ends(this_breakdown[[2]], "##") & 
-            this_breakdown[[3]] != "") {
-            this_breakdown[[3]] <- paste0("##", this_breakdown[[3]])
-        }
-        if (!stringr::str_starts(this_breakdown[[2]], "##") & 
-            this_breakdown[[1]] != "") {
-            this_breakdown[[1]] <- paste0(this_breakdown[[1]], "##")
-        }
-        current_breakdown[piece_to_work_on] <- list(this_breakdown)
-        current_breakdown <- unlist(current_breakdown)
+
+# .mp_tokenize_word_bidir -------------------------------------------------
+
+#' Tokenize a Word Bidirectionally
+#'
+#' Apply .mp_tokenize_word from both directions and pick the result with fewer
+#' pieces.
+#'
+#' @param word Word to tokenize.
+#' @param vocab Named integer vector containing vocabulary words. Should have
+#'   "vocab_split" attribute, with components named "prefixes", "words",
+#'   "suffixes".
+#'
+#' @return Input word as a list of tokens.
+#' @keywords internal
+.mp_tokenize_word_bidir <- function(word, vocab) {
+    vocab_split <- attr(vocab, "vocab_split")
+    t1 <- .mp_tokenize_word(word, vocab_split, dir = 1)
+    t2 <- .mp_tokenize_word(word, vocab_split, dir = -1)
+    if (length(t2) < length(t1) & length(t2) > 1) {
+        return(t2)
+    } else {
+        return(t1)
+    }
+}
+
+# .mp_tokenize_word_lookup -------------------------------------------------
+
+#' Tokenize a Word Including Lookup
+#'
+#' Look up a word in the table; go to fall-back otherwise.
+#'
+#' @inheritParams .mp_tokenize_word_bidir
+#' @param lookup A morphemepiece lookup table.
+#'
+#' @return Input word, broken into tokens.
+#' @keywords internal
+.mp_tokenize_word_lookup <- function(word, vocab, lookup) {
+    if (word %in% names(vocab)) { # punctuation, etc.
+        return(vocab[word])
     }
-    return(current_breakdown)
+    # may as well remove unbroken words from lookup, even though it's relatively
+    # small component?
```

Comment:
I don't grok what's happening here.

Reply:
The lookup table includes entries that don't get broken into pieces, like
`lookup["word"] = "word"`
but the way the code is currently structured, those words get looked up in the vocabulary (along with all the other non-alpha tokens, like punctuation). So there's really no reason to keep them in the lookup table. It's only about 5% of the lookup table right now, and it's not something to change in the code (at least not the code here).